cssparse/shared/src/main/scala/cssparse/CssParser.scala

```diff
@@ -0,0 +1,205 @@
+
+  val number_token = P( (CharIn("+-").? ~ ((digit.rep(1) ~ "." ~ digit.rep(1)) | digit.rep(1) | ("." ~ digit.rep(1) ~
+    (CharIn("eE") ~ CharIn("+-").? ~ digit.rep(1)).?))).! ).map(Ast.NumberToken)
+
+  val dimension_token = P( number_token.! ~ ident_token.! ) map
+    { case (number, ident) => Ast.DimensionToken(number, ident)}
+
+  val percentage_token = P( number_token.! ~ "%" ).map(Ast.PercentageToken)
+
+  val unicode_range_token = P( CharIn("Uu") ~ "+" ~
+    ((hex_digit.rep(min=1, max=6) ~ "?".rep(min=1, max=5)).! |
+    (hex_digit.rep(min=1, max=6).! ~ "-" ~ hex_digit.rep(min=1, max=6).!) |
+    hex_digit.rep(min=1, max=6).!)).map({
+      case hex: String => Ast.UnicodeRangeToken(hex, hex)
+      case (left: String, right: String) => Ast.UnicodeRangeToken(left, right)
+    })
+
+  val include_match_token =   P( "~=" ).map({_ => Ast.IncludeMatchToken()})
+  val dash_match_token =      P( "|=" ).map({_ => Ast.DashMatchToken()})
+  val prefix_match_token =    P( "^=" ).map({_ => Ast.PrefixMatchToken()})
+  val suffix_match_token =    P( "$=" ).map({_ => Ast.SuffixMatchToken()})
+  val substring_match_token = P( "*=" ).map({_ => Ast.SubstringMatchToken()})
+  val match_token = P( include_match_token | dash_match_token |
+                       prefix_match_token  | suffix_match_token |
+                       suffix_match_token  | substring_match_token |
+                       substring_match_token)
+
+  val column_token =          P( "||" ).map({_ => Ast.ColumnToken()})
+  val CDO_token =             P( "<!--" ).map({_ => Ast.CdoToken()})
+  val CDC_token =             P( "-->" ).map({_ => Ast.CdcToken()})
+
+  val delim_token = P( ("::" | CharIn("#$*+,-./:<>^~=!")).! ).map(Ast.DelimToken)
+
+  // any token except function_token
+  val simple_token: Parser[Option[Ast.SimpleToken]] = P(
+    whitespace_token    | at_word_token |
+    hash_token          | match_token |
+    column_token        | CDO_token |
+    CDC_token           | string_token |
+    unicode_range_token | percentage_token |
+    dimension_token     | url_token |
+    number_token        | ident_token |
+    delim_token ).map({
+      case st: Ast.SimpleToken => Some(st)
+      case _ => None
+    })
+
+  val bracketsBlock =       P( "(" ~ componentValue.rep ~ ")" ).map(values => Ast.BracketsBlock(values.flatten))
+  val curlyBracketsBlock =  P( "{" ~ componentValue.rep ~ "}" ).map(values => Ast.CurlyBracketsBlock(values.flatten))
+  val squareBracketsBlock = P( "[" ~ componentValue.rep ~ "]" ).map(values => Ast.SquareBracketsBlock(values.flatten))
+
+  val functionBlock = P( function_token ~ componentValue.rep ~ ")").map({
+    case (Ast.FunctionToken(name), values: Seq[Option[Ast.ComponentValue]]) =>
+      Ast.FunctionBlock(name, Ast.BracketsBlock(values.flatten))
+  })
+
+  val componentValue: Parser[Option[Ast.ComponentValue]] = P( simple_token | bracketsBlock |
+    curlyBracketsBlock | squareBracketsBlock | functionBlock ).map({
```

Comment:
Avoid pattern matching on `Any` like this. Call .map on the individual parsers to convert them to the same type before `|`ing them together